service/db.go

```diff
@@ -14,52 +14,93 @@ import (
 	"github.com/spacemeshos/poet/shared"
 )
 
+type dataType byte
+
+const (
+	typeProof dataType = iota
+	typeRound
+)
+
 var ErrNotFound = leveldb.ErrNotFound
 
-type ProofsDatabase struct {
+type ServiceDatabase struct {
 	db     *leveldb.DB
 	proofs <-chan shared.ProofMessage
+	rounds <-chan RoundData
+}
+
+func makeKey(t dataType, k []byte) []byte {
+	key := make([]byte, len(k)+1)
+	key = append(key, byte(t))
+	return append(key, k...)
 }
 
-func (db *ProofsDatabase) Get(ctx context.Context, roundID string) (*shared.ProofMessage, error) {
-	data, err := db.db.Get([]byte(roundID), nil)
+func (db *ServiceDatabase) GetProof(ctx context.Context, roundID string) (*shared.ProofMessage, error) {
+	data, err := db.db.Get(makeKey(typeProof, []byte(roundID)), nil)
 	if err != nil {
-		return nil, fmt.Errorf("get proof for %s from DB: %w", roundID, err)
+		return nil, fmt.Errorf("getting proof for round %s from DB: %w", roundID, err)
 	}
 
 	proof := &shared.ProofMessage{}
 	if _, err := proof.DecodeScale(scale.NewDecoder(bytes.NewReader(data))); err != nil {
-		return nil, fmt.Errorf("failed to get deserialize proof: %w", err)
+		return nil, fmt.Errorf("deserializing proof for round %s: %w", roundID, err)
 	}
 	return proof, nil
 }
 
-func NewProofsDatabase(dbPath string, proofs <-chan shared.ProofMessage) (*ProofsDatabase, error) {
+func (db *ServiceDatabase) GetRoundMembers(ctx context.Context, roundID string) ([][]byte, error) {
+	data, err := db.db.Get(makeKey(typeRound, []byte(roundID)), nil)
+	if err != nil {
+		return nil, fmt.Errorf("getting members for round %s from DB: %w", roundID, err)
+	}
+
+	if members, _, err := scale.DecodeSliceOfByteSlice(scale.NewDecoder(bytes.NewReader(data))); err != nil {
+		return nil, fmt.Errorf("deserializing round %s members: %w", roundID, err)
+	} else {
+		return members, nil
+	}
+}
+
+func NewServiceDatabase(
+	dbPath string,
+	proofs <-chan shared.ProofMessage,
+	rounds <-chan RoundData,
+) (*ServiceDatabase, error) {
 	db, err := leveldb.OpenFile(dbPath, nil)
 	if err != nil {
-		return nil, fmt.Errorf("failed to open database @ %s: %w", dbPath, err)
+		return nil, fmt.Errorf("opening database @ %s: %w", dbPath, err)
 	}
 
-	return &ProofsDatabase{db, proofs}, nil
+	return &ServiceDatabase{db, proofs, rounds}, nil
 }
 
-func (db *ProofsDatabase) Run(ctx context.Context) error {
+func (db *ServiceDatabase) Run(ctx context.Context) error {
```

Comment:
I feel like the service does more than it should. A database persists and fetches the entities from the store, but imo shouldn't have a `Run` method that kicks off an endless loop for processing data.

Suggestion: instead of having this loop here (that needs to be executed on its own go-routine), how about if the writers to the channels `db.proofs` and `db.rounds` just call `AddProof` and `AddRound` on the `ServiceDB`? Then no endless loop running on its own routine synchronising between channels is needed.

LevelDB is according to its documentation even save to be used concurrently, so the whole synchronization to avoid calling `Put` concurrently isn't even necessary (and if it would I would prefer just using a mutex in `ServiceDatabase`).

This would also allow to simplify the tests in `db_test.go`.

Reply:
After discussing with @poszu we can leave the `ServiceDB` as it is and might take a look at it again in the future.